cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs

```diff
@@ -353,70 +373,137 @@ readSigningKeyFile skFile =
+        return $ makeShelleyBootstrapWitness nw txbody skByron
+      (AShelleyKeyWitness skShelley, _) ->
+        return $ makeShelleyKeyWitness txbody skShelley
+      (AShelleyScriptWitness scShelley, _) ->
+        return $ makeShelleyScriptWitness (makeMultiSigScript scShelley)
 
   firstExceptT ShelleyTxCmdWriteFileError
     . newExceptT
     $ Api.writeFileTextEnvelope oFile Nothing witness
 
-runTxSignWitness :: TxBodyFile -> [WitnessFile] -> OutputFile -> ExceptT ShelleyTxCmdError IO ()
+runTxSignWitness
+  :: TxBodyFile
+  -> [WitnessFile]
+  -> OutputFile
+  -> ExceptT ShelleyTxCmdError IO ()
 runTxSignWitness (TxBodyFile txBodyFile) witnessFiles (OutputFile oFp) = do
     txBody <- firstExceptT ShelleyTxCmdReadTextViewFileError
       . newExceptT
       $ Api.readFileTextEnvelope Api.AsShelleyTxBody txBodyFile
-    witnesses <- firstExceptT ShelleyTxCmdReadTextViewFileError
-      $ mapM readWitnessFile witnessFiles
+    witnesses <- mapM readWitnessFile witnessFiles
     let tx = Api.makeSignedTransaction witnesses txBody
     firstExceptT ShelleyTxCmdWriteFileError
       . newExceptT
       $ Api.writeFileTextEnvelope oFp Nothing tx
-  where
-    readWitnessFile
-      :: WitnessFile
-      -> ExceptT (Api.FileError Api.TextEnvelopeError) IO (Witness Shelley)
-    readWitnessFile (WitnessFile fp) =
-      newExceptT (Api.readFileTextEnvelope AsShelleyWitness fp)
+
+readWitnessFile :: WitnessFile -> ExceptT ShelleyTxCmdError IO (Witness Shelley)
+readWitnessFile (WitnessFile fp) =
+  firstExceptT ShelleyTxCmdReadTextViewFileError $ newExceptT (Api.readFileTextEnvelope AsShelleyWitness fp)
+
+runTxBuildMultiSig :: MultiSigScriptObject -> Maybe OutputFile -> ExceptT ShelleyTxCmdError IO ()
+runTxBuildMultiSig msso mOutputFile = do
+  ms <- convertToMultiSig msso
+  case mOutputFile of
+    Just (OutputFile outFp) -> liftIO $ LBS.writeFile outFp $ encodePretty ms
+    Nothing -> liftIO . C8.putStrLn $ encodePretty ms
+ where
+  readPaymentVerificationKeys :: [VerificationKeyFile] -> ExceptT ShelleyTxCmdError IO [VerificationKey PaymentKey]
+  readPaymentVerificationKeys fps = do
+    eVerKeys <- liftIO $ mapM (readFileTextEnvelopeAnyOf fileTypes . unVerificationKeyFile) fps
+    someKeys <- sequence $ map (firstExceptT ShelleyTxCmdReadTextViewFileError . hoistEither) eVerKeys
+    right $ map convertToVerificationKey someKeys
+
+  fileTypes = [ FromSomeType (AsVerificationKey AsPaymentKey) SomePaymentVerificationKey
+              , FromSomeType (AsVerificationKey AsGenesisUTxOKey) SomeGenesisUTxOVerificationKey
+              ]
+
+  convertToVerificationKey :: SomeVerificationKey -> VerificationKey PaymentKey
+  convertToVerificationKey sVkey =
+    case sVkey of
+      SomePaymentVerificationKey vK -> vK
+      SomeGenesisUTxOVerificationKey utxoKey -> castVerificationKey utxoKey
+
+  convertToMultiSig :: MultiSigScriptObject -> ExceptT ShelleyTxCmdError IO MultiSigScript
+  convertToMultiSig so =
```

Comment:
The impl here would change if we did either 1. remove specifying scripts on the cli at all, or 2. made the cli form properly follow the structure of the multisig script language.

Reply:
I think it would be useful to create multisig scripts from the cli. @disassembler are you using the cli to create multisig scripts?